### ISSUE TITLE
Compositing material model does not allow for the underlying material models to provide additonal output

### ISSUE THREAD

**Issue by anne-glerum** [geodynamics/aspect#3161]
The `visco plastic` material model fills several additional outputs, e.g. `plastic_yielding`. When using the `compositing` material model to refer to the `visco_plastic` model to compute the viscosity, these outputs are not filled. This came up when asking for the `named additional outputs` visualization postprocessor. 

Although there is no guarantee that the components requested from the underlying material model correspond to the additional output that this material model computes (think `thermal_conductivity` versus `plastic_yielding`), I think in principle it should be an option to ask the underlying material models for their additional output when using the `compositing` material model.

**Comment by gassmoeller** [comment]
That seems indeed wrong. Is the problem that it is not possible to ask for the additional outputs, or is the problem that they will never be created in the first place (because `Compositing` does not forward the call to `create_additional_named_outputs` to the base models)? Because if they are created they should also be available from `MaterialModelOutputs`.

**Comment by anne-glerum** [comment]
`Compositing` does not forward the `create_additional_named_outputs` call to the base models, nor does it do anything else with AdditionalOutput. 

The error message when activating the pp:
`Exception 'ExcMessage("You activated the postprocessor <named additional outputs>, but there are no additional outputs " "provided by the material model. Either remove the postprocessor, or check why no output is provided.")' on rank 0 on processing: 
`
`An error occurred in line <68> of file </Users/ACG/Documents/PhD/Aspect/source/25-05-2019/aspect/source/postprocess/visualization/named_additional_outputs.cc> in function
    virtual void aspect::Postprocess::VisualizationPostprocessors::NamedAdditionalOutputs<2>::initialize() [dim = 2]
The violated condition was: 
    out.additional_outputs.size() > 0
Additional information: 
    You activated the postprocessor <named additional outputs>, but there are no additional outputs provided by the material model. Either remove the postprocessor, or check why no output is provided.`

**Comment by gassmoeller** [comment]
> Compositing does not forward the create_additional_named_outputs call to the base models, nor does it do anything else with AdditionalOutput.

Then it might be enough to create such a function in `Compositing` and let it call the corresponding functions of all the base models it owns?

**Comment by anne-glerum** [comment]
> Then it might be enough to create such a function in `Compositing` and let it call the corresponding functions of all the base models it owns?

I don't think this is enough because `Compositing` calls the `evaluate` function with a different `MaterialModelOutputs` than it itself is called. Therefore just forwarding the `create_additional_named_outputs` function to the base models does not get the outputs filled. 

When calling the create function on the base `MaterialModelOutputs` before evaluating the base models, these outputs get filled, but then I'm not sure how to get them into the outputs of the `Compositing` model.

**Comment by gassmoeller** [comment]
Ah, I see the problem. The one solution I can think of at the moment is to duplicate the loop over all base models in `Compositing<dim>::evaluate`. First evaluate all models on `out` to generate the additional outputs, then evaluate all models on `base_output` and copy the required properties into `out`. Inefficient, but it should work.

There is one more complication: Base models might behave differently depending on whether additional outputs are there or not. So we really need to generate the additional outputs for the `base_out` as well. I think you can make `base_out` a member of the `Compositing` class, and then create a `create_additional_named_outputs` function that  first copies `base_out = out`, and then creates all additional outputs for both of them.

**Comment by anne-glerum** [comment]
Update: #3161 creates additional output generated by the base material models for the incoming output and for the base_output that is used for the properties queried from the base models. It does however not cover additional output generated elsewhere by for example the NewtonHandler. The base models will therefore not get the elsewhere generated output when being evaluated for the regular output, but will have it when being evaluated for the additional output.
